sycl/test-e2e/KernelCompiler/kernel_compiler_sycl.cpp

```diff
@@ -0,0 +1,167 @@
+
+  // intentionally using deprecated routine, as opposed to this_work_item::get_nd_item<1>()
+  sycl::nd_item<1> Item = sycl::ext::oneapi::experimental::this_nd_item<1>();
+
+  sycl::id<1> GId = Item.get_global_id();
+  ptr[GId.get(0)] = AddEm(GId.get(0), 37);
+}
+)===";
+
+auto constexpr BadSource = R"===(
+#include <sycl/sycl.hpp>
+
+extern "C" SYCL_EXTERNAL SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((sycl::ext::oneapi::experimental::nd_range_kernel<1>))
+void ff_cp(int *ptr) {
+
+  sycl::nd_item<1> Item = sycl::ext::oneapi::this_work_item::get_nd_item<1>();
+
+  sycl::id<1> GId = Item.get_global_id() + no semi colon !!
+  ptr[GId.get(0)] = GId.get(0) + 41;
+}
+)===";
+
+void test_1(sycl::queue &Queue, sycl::kernel &Kernel) {
+  constexpr int Range = 10;
+  int *usmPtr = sycl::malloc_shared<int>(Range, Queue);
+  int start = 3;
+
+  sycl::nd_range<1> R1{{Range}, {1}};
+
+  bool Passa = true;
+
+  memset(usmPtr, 0, Range * sizeof(int));
+  Queue.submit([&](sycl::handler &Handler) {
+    Handler.set_arg(0, usmPtr);
+    // Handler.set_arg(1, start);
+    // Handler.set_arg(2, Range);
+    Handler.parallel_for(R1, Kernel);
+  });
+  Queue.wait();
+
+  for (int i = 0; i < Range; i++) {
+    std::cout << usmPtr[i] << " ";
+    // assert(usmPtr[i] = i + 42);
+  }
+  std::cout << std::endl;
+
+  sycl::free(usmPtr, Queue);
+}
+
+void test_build_and_run() {
+  namespace syclex = sycl::ext::oneapi::experimental;
+  using source_kb = sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source>;
+  using exe_kb = sycl::kernel_bundle<sycl::bundle_state::executable>;
+
+  sycl::queue q;
+  sycl::context ctx = q.get_context();
+
+  bool ok =
+      q.get_device().ext_oneapi_can_compile(syclex::source_language::sycl);
+  if (!ok) {
+    std::cout << "Apparently this device does not support SYCL source "
+                 "kernel bundle extension: "
+              << q.get_device().get_info<sycl::info::device::name>()
+              << std::endl;
+    return;
+  }
+
+  // create from source
+  source_kb kbSrc = syclex::create_kernel_bundle_from_source(
+      ctx, syclex::source_language::sycl, SYCLSource,
+      syclex::properties{syclex::include_files{"AddEm.h", AddEmH}});
+
+  // double check kernel_bundle.get_source() / get_backend()
+  sycl::context ctxRes = kbSrc.get_context();
+  assert(ctxRes == ctx);
+  sycl::backend beRes = kbSrc.get_backend();
+  assert(beRes == ctx.get_backend());
+
+  // compilation of empty prop list, no devices
+  exe_kb kbExe1 = syclex::build(kbSrc);
+
+  // compilation with props and devices
+  std::string log;
+  std::vector<std::string> flags{"-g", "-fno-fast-math"};
+  std::vector<sycl::device> devs = kbSrc.get_devices();
+  exe_kb kbExe2 = syclex::build(
+      kbSrc, devs,
+      syclex::properties{syclex::build_options{flags}, syclex::save_log{&log}});
+  assert(log.find("warning: 'this_nd_item<1>' is deprecated") !=
+         std::string::npos);
+
+  sycl::kernel k = kbExe2.ext_oneapi_get_kernel("__sycl_kernel_ff_cp");
```

Comment:
According to the spec, you should be able to do this:

```suggestion
  sycl::kernel k = kbExe2.ext_oneapi_get_kernel("ff_cp");
```

I do not think this requires any help from the front-end for an `extern "C"` kernel.  The metadata from the front-end only helps when the application uses `registered_kernel_names`, but you don't need to use that property for `extern "C"` kernels.

My expectation was that the SYCL runtime algorithm for `ext_oneapi_get_kernel` would be like this:

* Is the name in the `registered_kernel_names` list?
* If yes: use the metadata to find the mangled kernel name
* If no: assume the name is for an `extern "C"` kernel and try to find a kernel named `"__sycl_kernel_"` + `name`.

Reply:
I altered the logic slightly, but it's done. `get_kernel("ff_cp")` works, as does using the full mangled name (if you know it) until we get that support in. Updated the test as well.